// @ionic/cloud/dist/es5/core.js
"use strict";
var Core = (function () {
    function Core(deps, cfg) {
        this.cfg = cfg;
        this._version = '0.8.0-beta.11';
        this.config = deps.config;
        this.logger = deps.logger;
        this.emitter = deps.emitter;
        this.insights = deps.insights;
        this.registerEventHandlers();
        this.insights.track('mobileapp.opened');
    }
    Object.defineProperty(Core.prototype, "version", {
        get: function () {
            return this._version;
        },
        enumerable: true,
        configurable: true
    });
    Core.prototype.registerEventHandlers = function () {
        var _this = this;
        this.emitter.on('cordova:resume', function (data) {
            _this.insights.track('mobileapp.opened');
        });
        this.emitter.on('push:notification', function (data) {
            if (data.message.app.asleep || data.message.app.closed) {
                _this.insights.track('mobileapp.opened.push');
            }
        });
    };
    return Core;
}());
exports.Core = Core;
